### PR TITLE
Make rustdoc lints a tool lint instead of built-in

### LINKED ISSUE #78786: Rustdoc should never silence lints from the `rustdoc` lint group
I think the right fix here is to never silence lints from the `rustdoc` lint group, then there wouldn't be so many places to add new lints.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/pull/78114#issuecomment-722542923_

### PR DESCRIPTION
**Pull request by jyn514**
- Rename `broken_intra_doc_links` to `rustdoc::broken_intra_doc_links` (and similar for other rustdoc lints; I don't expect any others to be used frequently, though).
- Ensure that the old lint names still work and give deprecation errors
- Register lints even when running doctests
- Move lint machinery into a separate file
- Add `declare_rustdoc_lint!` macro

Unblocks https://github.com/rust-lang/rust/pull/80300, https://github.com/rust-lang/rust/pull/79816, https://github.com/rust-lang/rust/pull/80965. Makes the strangeness in https://github.com/rust-lang/rust/pull/77364 more apparent to the end user (note that `missing_docs` is *not* moved to rustdoc in this PR). Closes https://github.com/rust-lang/rust/issues/78786.

## Current status

This is blocked on #82620 (see https://github.com/rust-lang/rust/pull/80527#issuecomment-787401519)

### CODE REVIEW

**Comment by jyn514** [comment]
@rfcbot fcp merge

This changes the following lints to have a `rustdoc::` prefix:

- broken_intra_doc_links
- private_intra_doc_links
- missing_doc_code_examples
- private_doc_tests
- invalid_codeblock_attributes
- invalid_html_tags
- non_autolinks

The old names warn about changing to use the `rustdoc::`, but otherwise do not have an effect (and that's tested: https://github.com/rust-lang/rust/pull/80527/files#diff-1c06cd7b195eef73c25529b9549b2ed5ce4db5228155ee23c28169aff8c5fb35).

## Potential alternatives

- Don't ever change the names. However, several compiler team members have expressed dissatisfaction with the status quo, and I personally think that having a `rustdoc::` prefix makes it more clear that the lint will only appear when running rustdoc, not the compiler directly.

**Comment by scottmcm** [comment]
Makes sense to have the rustdoc lints be scoped to `rustdoc::`, like the clippy ones are scoped now.

@rfcbot reviewed

**Comment by ehuss** [comment]
Are you planning to update the documentation?

Also, there seems to be a stray modification to `Cargo.lock`.

**Comment by jyn514** [comment]
> Are you planning to update the documentation?

What documentation should I update?

> Also, there seems to be a stray modification to Cargo.lock.

Oops, fixed, thanks.

**Comment by ehuss** [comment]
This page should probably talk about the `rustdoc::` prefix: https://github.com/rust-lang/rust/blob/master/src/doc/rustdoc/src/lints.md

It should probably be careful on how it handles MISSING_DOCS.  That is already documented in the [rustc book](https://doc.rust-lang.org/nightly/rustc/lints/listing/allowed-by-default.html#missing-docs), so you could maybe just consider changing it to link to that.  Or you could just call out that it doesn't need the `rustdoc::` prefix.

Also, I'm uncertain why MISSING_CRATE_LEVEL_DOCS is not being moved.  It is not processed by rustc, it is a rustdoc-only lint AFAIK. It seems a little strange to me to have that be the only rustdoc-only lint not prefixed with `rustdoc::`.

**Comment by jyn514** [comment]
> Also, I'm uncertain why MISSING_CRATE_LEVEL_DOCS is not being moved. It is not processed by rustc, it is a rustdoc-only lint AFAIK. It seems a little strange to me to have that be the only rustdoc-only lint not prefixed with rustdoc::.

Oops, good catch! I changed it to also have a rustdoc:: prefix.

> This page should probably talk about the rustdoc:: prefix: https://github.com/rust-lang/rust/blob/master/src/doc/rustdoc/src/lints.md

Done.

> It should probably be careful on how it handles MISSING_DOCS. That is already documented in the rustc book, so you could maybe just consider changing it to link to that. Or you could just call out that it doesn't need the rustdoc:: prefix.

I called out that it doesn't need the rustdoc:: prefix. Rustdoc silences all lints by default, so since MISSING_DOCS isn't silenced, I think it makes sense to mention it in the docs.

File: src/test/rustdoc-ui/check.stderr
Comment (jyn514) [inline src/test/rustdoc-ui/check.stderr:33]: @ehuss I found why I forgot to update this earlier: it wasn't part of the `rustdoc` lint group.

File: compiler/rustc_lint/src/lib.rs
Comment (GuillaumeGomez) [inline compiler/rustc_lint/src/lib.rs:327]: `register_removed` you mean? ;)

Comment (jyn514) [reply compiler/rustc_lint/src/lib.rs]: Thanks, fixed.

File: src/doc/rustdoc/src/lints.md
Comment (jyn514) [inline src/doc/rustdoc/src/lints.md:285]: As @joshtriplett noticed, the output was incorrect before: `invalid_html_tags` is allowed by default. This happened because doctests `lints.md` are not tested. He's going to open an issue for testing this output.

Comment (joshtriplett) [inline src/doc/rustdoc/src/lints.md:285]: Note based on discussion with @jyn514 on Zulip: despite the output change here, the lint was not previously enabled by default. The old output just wasn't accurate for the current state of rustc.

I've filed https://github.com/rust-lang/rust/issues/81305 about getting these files checked as doctests. I don't consider that a blocker for this PR, though.

**Review by joshtriplett — APPROVED** [review]

**Comment by joshtriplett** [comment]
@rfcbot reviewed

**Comment by jyn514** [comment]
ping @Manishearth @cramertj @nikomatsakis @ollie27 @pnkfelix @withoutboats: https://github.com/rust-lang/rust/pull/80527#issuecomment-761934254 is waiting on your review.

**Comment by Manishearth** [comment]
Oops, thought I'd checkboxed this

**Comment by jyn514** [comment]
ping @nikomatsakis @ollie27 @pnkfelix @withoutboats: https://github.com/rust-lang/rust/pull/80527#issuecomment-761934254 is waiting on your review.

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

**Comment by nikomatsakis** [comment]
I also checked for @withoutboats, given that they've withdrawn from the teams.

**Comment by jyn514** [comment]
@GuillaumeGomez do you want to do the honors? :)

**Comment by GuillaumeGomez** [comment]
Happy to! Thanks everyone for your input and thanks @jyn514 for working on this!

@bors: r+

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/82574#issuecomment-786974987)

@bors r-

**Comment by camelid** [comment]
Part of the log:

```
thread 'doc::short_message_format' panicked at '
Expected: execs
    but: exited with Some(0)
--- stdout

--- stderr
 Documenting foo v0.0.1 (/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t715/foo)
src/lib.rs:2:9: warning: lint `broken_intra_doc_links` has been renamed to `rustdoc::broken_intra_doc_links`
src/lib.rs:4:6: warning: unresolved link to `bad_link`
warning: 2 warnings emitted
    Finished dev [unoptimized + debuginfo] target(s) in 1.17s
', src/tools/cargo/crates/cargo-test-support/src/lib.rs:739:13


failures:
    doc::doc_message_format
    doc::short_message_format
```

**Comment by jyn514** [comment]
@ehuss do you have suggestions for how to fix this error? Normally I would make a separate PR fixing the issue, then update the cargo submodule as part of #80527, but it's not possible to fix this separately because the new lint names don't exist yet:
```
error[E0710]: an unknown tool name found in scoped lint: `rustdoc::broken_intra_doc_links`
 --> lint.rs:1:9
  |
1 | #![deny(rustdoc::broken_intra_doc_links)]
  |         ^^^^^^^
```
I guess I could disable the test altogether temporarily - would you be ok with that?

**Comment by ehuss** [comment]
I'm a bit confused by something.  The allow/warn/deny/forbid attributes do not seem to work with renamed lints, which is not the behavior I would expect.  I wouldn't expect either of those tests to fail if the lint is just renamed.  Does that match your intuition?

I guess that is not directly related to this PR, and a bit unfortunate that unknown tool names aren't just a warning. I imagine that will make the transition a bit difficult for people, since it will be impossible to use something that works with both versions.  

If you want to post a PR that marks those two tests with `#[ignore]`, I can approve that and do a quick submodule update, and then once this lands on nightly, they can be switched to nightly-only tests.

**Comment by jyn514** [comment]
> I'm a bit confused by something. The allow/warn/deny/forbid attributes do not seem to work with renamed lints, which is not the behavior I would expect. I wouldn't expect either of those tests to fail if the lint is just renamed. Does that match your intuition?

Oh huh, you're right. I thought I added a test for this ... I'll add one.

```
warning: lint `broken_intra_doc_links` has been renamed to `rustdoc::broken_intra_doc_links`
 --> lint.rs:1:9
  |
1 | #![deny(broken_intra_doc_links)]
  |         ^^^^^^^^^^^^^^^^^^^^^^ help: use the new name: `rustdoc::broken_intra_doc_links`
  |
  = note: `#[warn(renamed_and_removed_lints)]` on by default

warning: unresolved link to `x`
 --> lint.rs:2:6
  |
2 | //! [x]
  |      ^ no item named `x` in scope
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```

Do you know why this isn't handled by the lint framework? I'd expect `register_renamed` to apply the lint for both the old and new name.

**Comment by jyn514** [comment]
I tried removing `register_renamed` when running rustc:
```diff
diff --git a/compiler/rustc_lint/src/lib.rs b/compiler/rustc_lint/src/lib.rs
index 547779dd685..d56c2589c9c 100644
--- a/compiler/rustc_lint/src/lib.rs
+++ b/compiler/rustc_lint/src/lib.rs
@@ -336,6 +336,8 @@ fn register_builtins(store: &mut LintStore, no_interleave_lints: bool) {
         "invalid_html_tags",
         "non_autolinks",
     ];
+
+    /*
     for rustdoc_lint in RUSTDOC_LINTS {
         store.register_removed(rustdoc_lint, &format!("use `rustdoc::{}` instead", rustdoc_lint));
     }
@@ -343,6 +345,7 @@ fn register_builtins(store: &mut LintStore, no_interleave_lints: bool) {
         "intra_doc_link_resolution_failure",
         "use `rustdoc::broken_intra_doc_links` instead",
     );
+    */
 
     store.register_removed("unknown_features", "replaced by an error");
     store.register_removed("unsigned_negation", "replaced by negate_unsigned feature gate");
```
but it didn't help.

**Comment by ehuss** [comment]
Filed #82615 to track the issue with renaming not working as expected.  Looks like it regressed in 1.21.

**Comment by jyn514** [comment]
I opened https://github.com/rust-lang/rust/pull/82620 fixing the issue and rebased this over it. I confirmed locally that fixes the cargo tests.

@rustbot label: +S-blocked -S-waiting-on-author

**Comment by camelid** [comment]
#82620 has merged! @rustbot label: -S-blocked +S-waiting-on-author

**Comment by camelid** [comment]
Same unrelated failure as https://github.com/rust-lang/rust/pull/82744#issuecomment-790283763.

**Comment by lopopolo** [comment]
I would have liked to see the `rustdoc` lint group exist on stable before this change was merged.

It is not possible to run both nightly and stable builds with `-D warnings` and `warn(broken_intra_doc_links)`/`warn(rustdoc:: broken_intra_doc_links)` now because, despite `allow(unknown_lints)`, the unknown tool name results in a hard error:

```
error[E0710]: an unknown tool name found in scoped lint: `rustdoc::broken_intra_doc_links`
 --> artichoke-core/src/lib.rs:9:9
  |
9 | #![warn(rustdoc::broken_intra_doc_links)]
  |         ^^^^^^^
```

**Comment by jyn514** [comment]
> It is not possible to run both nightly and stable builds with -D warnings and warn(broken_intra_doc_links)/warn(rustdoc:: broken_intra_doc_links) now because, despite allow(unknown_lints), the unknown tool name results in a hard error:

If you use the original name (`broken_intra_doc_links)` and `allow(renamed_removed_lints)`, it should work.

**Comment by lopopolo** [comment]
@jyn514 does the introduction of a new tool name mean that declaring this lint as a crate level warn pragma will bump the MSRV of a crate?

**Comment by lopopolo** [comment]
> If you use the original name (`broken_intra_doc_links`) and `allow(renamed_removed_lints)`, it should work.

Since the lint has been removed, does this mean running nightly rustdoc will not detect broken intradoc links anymore?

**Comment by jyn514** [comment]
> Since the lint has been removed, does this mean running nightly rustdoc will not detect broken intradoc links anymore?

No, the lint name will still apply (since https://github.com/rust-lang/rust/pull/82620.

> @jyn514 does the introduction of a new tool name mean that declaring this lint as a crate level warn pragma will bump the MSRV of the crate?

Yes, unfortunately. There's no way to fix this because past releases are immutable (ideally the unknown tool name would only be a warning and not a hard error).

**Comment by lopopolo** [comment]
> ideally the unknown tool name would only be a warning and not a hard error

Has this change landed in nightly/is there a ticket for it?

**Comment by jyn514** [comment]
> Has this change landed in nightly/is there a ticket for it?

I think https://github.com/rust-lang/rust/issues/66079#issuecomment-788589193 is the closest issue.

**Comment by lopopolo** [comment]
Thank you!

**Comment by jyn514** [comment]
> I would have liked to see the rustdoc lint group/tool name exist on stable before this change was merged.

@lopopolo I opened https://github.com/rust-lang/rust/pull/83203 to turn off the warnings until this makes it to stable.

**Pull request merged**
